### PR TITLE
Serialize both AssemblyNameExtension and SystemState, then opt into their use for all on-disk RAR caching

### PR DESCRIPTION
**Pull request by Forgind**
Please merge. I worked hard on the commits. :)

This is based off of #5514. That needs to go in first.

### CODE REVIEW

File: src/Tasks/SystemState.cs [hunk 1]
```diff
@@ -540,16 +676,93 @@ out fileState.frameworkName
...
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = stateFiles.Length > 0;
+            HashSet<string> assembliesFound = new HashSet<string>();
+
+            foreach (string stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                var deserializeOptions = new JsonSerializerOptions() { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
+                deserializeOptions.Converters.Add(new SystemState.Converter());
+                SystemState sysBase = JsonSerializer.Deserialize<SystemState>(File.ReadAllText(stateFile), deserializeOptions);
```
Comment (cdmihai) [inline src/Tasks/SystemState.cs:698]: Is json really faster than binary serialization? I thought the whole point of this feature is to maximize perf. Can you please do some measurements on real SystemState data to compare binary serialization vs json, so we're aware what the impact is?

Comment (Forgind) [reply src/Tasks/SystemState.cs:698]: Most of the reason is that BinaryFormatter is unsafe, and we're supposed to move all of MSBuild off of it. I should measure perf at some point, but I haven't gotten around to it yet.

Comment (cdmihai) [reply src/Tasks/SystemState.cs:698]: If that's the case, we should be replacing it with another binary serialization library that's safe and as fast or faster.

Comment (Forgind) [reply src/Tasks/SystemState.cs]: I just did a janky test comparing the time of:
```
[Fact]
        public void SpeedTest()
        {
            string data = File.ReadAllText(@"<path>");
            for (int i = 0; i < 1000; i++)
            {
                using (FileStream s = new FileStream(@"<path>", FileMode.Open))
                {
                    var formatter = new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();
                    SystemState t = formatter.Deserialize(s) as SystemState;
                }
            }
        }
```
minus the time of
```
[Fact]
        public void SpeedTest()
        {
            for (int i = 0; i < 1000; i++)
            {
                string data = File.ReadAllText(@"<path>");
            }
        }
```

to the time of

```
[Fact]
        public void TestSpeed()
        {
            string data = File.ReadAllText(@"<path>");
            var deserializeOptions = new JsonSerializerOptions() { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
            deserializeOptions.Converters.Add(new SystemState.Converter());
            for (int i = 0; i < 1000; i++)
                JsonSerializer.Deserialize<SystemState>(data, deserializeOptions);
        }
```

For the json serializer, I used a statefile created by combining the state files cached when building all of MSBuild, whereas I used just the state file cache for Microsoft.Build.Engine.UnitTests for BinaryFormatter, so it should be smaller. Still, the two runs for the binary formatter took 2 seconds (unfortunately not more precise) and 435 ms, whereas json took 996 ms, suggesting that for this test at least, json was slightly faster.

File: src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs [hunk 2]
```diff
@@ -1855,27 +1857,46 @@ private void LogConflict(Reference reference, string fusionName)
...
         {
-            if (!string.IsNullOrEmpty(_stateFile) && _cache.IsDirty)
+            if (!string.IsNullOrEmpty(AssemblyInformationCacheOutputPath))
+            {
+                _cache.SerializePrecomputedCache(AssemblyInformationCacheOutputPath, Log, calculateMvid);
+            }
+            else if (!string.IsNullOrEmpty(_stateFile) && _cache.IsDirty)
             {
-                _cache.SerializeCache(_stateFile, Log);
+                var deserializeOptions = new JsonSerializerOptions() { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping };
```
Comment (cdmihai) [inline src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:1897]: Why not hide the JSON format implementation detail under SystemState.<Serialize|Deserialize>Cache

Comment (Forgind) [reply src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs:1897]: I don't implement JsonSerializer directly, unfortunately. As I recall, that's what determines how the reader/writer determine what counts as an interesting token. I just implement the API they provide with specifics, and they do the serializing.

File: src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                t._cache.instanceLocalFileStateCache = new Dictionary<string, SystemState.FileState>() {
+                    { Path.Combine(standardCache.Path, "assembly1"), new SystemState.FileState(DateTime.Now) },
+                    { Path.Combine(standardCache.Path, "assembly2"), new SystemState.FileState(DateTime.Now) { Assembly = new Shared.AssemblyNameExtension("hi") } } };
+                t.StateFile = standardCache.Path;
+                t._cache.IsDirty = true;
+                t.WriteStateFile(calculateMvid);
+
+                string dllName = Path.Combine(Path.GetDirectoryName(standardCache.Path), "randomFolder", "dll.dll");
+                t._cache.instanceLocalFileStateCache.Add(dllName,
+                    new SystemState.FileState(DateTime.Now) {
```
Comment (cdmihai) [inline src/Tasks.UnitTests/RARPrecomputedCache_Tests.cs:72]: Sanity check: This is sufficient data to hit the Converter from `AssemblyNameExtension` right?

**Review by cdmihai — APPROVED** [review]
LGTM, as long as the perf results from Orchard Core look good 👍

**Pull request closed without merging**